[netatmo] Handler initialization fails (duplicate element)

I upgraded a 3.4.0 instance to 4.0.0 SNAPSHOT today and found:

```
15:57:14.040 [INFO ] [hab.event.ThingStatusInfoChangedEvent] - Thing 'netatmo:home:383aed4627:6022fd77c380915550359cfd' changed from UNINITIALIZED (NOT_YET_READY) to UNINITIALIZED (HANDLER_REGISTERING_ERROR): duplicate element: 
15:57:14.026 [ERROR] [.core.thing.internal.ThingManagerImpl] - Exception occurred while calling thing handler factory 'org.openhab.binding.netatmo.internal.NetatmoHandlerFactory@46fc12bb': duplicate element: 
java.lang.IllegalArgumentException: duplicate element: 
	at java.util.ImmutableCollections$SetN.<init>(ImmutableCollections.java:918) ~[?:?]
	at java.util.Set.of(Set.java:505) ~[?:?]
	at org.openhab.binding.netatmo.internal.handler.capability.HomeCapability.initialize(HomeCapability.java:61) ~[?:?]
	at org.openhab.binding.netatmo.internal.handler.capability.CapabilityMap.put(CapabilityMap.java:34) ~[?:?]
	at org.openhab.binding.netatmo.internal.NetatmoHandlerFactory.lambda$4(NetatmoHandlerFactory.java:148) ~[?:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at org.openhab.binding.netatmo.internal.NetatmoHandlerFactory.buildHandler(NetatmoHandlerFactory.java:120) ~[?:?]
	at org.openhab.binding.netatmo.internal.NetatmoHandlerFactory.lambda$2(NetatmoHandlerFactory.java:107) ~[?:?]
	at java.util.Optional.map(Optional.java:260) ~[?:?]
	at org.openhab.binding.netatmo.internal.NetatmoHandlerFactory.createHandler(NetatmoHandlerFactory.java:107) ~[?:?]
	at org.openhab.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:129) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.doRegisterHandler(ThingManagerImpl.java:528) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.registerHandler(ThingManagerImpl.java:516) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.registerAndInitializeHandler(ThingManagerImpl.java:924) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.checkMissingPrerequisites(ThingManagerImpl.java:1120) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

```
UID: netatmo:home:383aed4627:<id>
label: Netatmo Home
thingTypeUID: netatmo:home
configuration:
  refreshInterval: 180
  id: <id>
bridgeUID: netatmo:account:383aed4627
```

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/CommonInterface.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/CommonInterface.java
@@ -122,9 +122,11 @@ public interface CommonInterface {
     }
 
     default Optional<CommonInterface> recurseUpToHomeHandler(@Nullable CommonInterface handler) {
-        return handler == null ? Optional.empty()
-                : handler.getCapabilities().get(HomeCapability.class).isPresent() ? Optional.of(handler)
-                        : recurseUpToHomeHandler(handler.getBridgeHandler());
+        if (handler == null) {
+            return Optional.empty();
+        }
+        return handler.getCapabilities().get(HomeCapability.class).isPresent() ? Optional.of(handler)
+                : recurseUpToHomeHandler(handler.getBridgeHandler());
     }
 
     default List<CommonInterface> getActiveChildren() {

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/CommonInterface.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/CommonInterface.java
@@ -121,13 +121,10 @@ public interface CommonInterface {
                 .filter(channel -> ChannelKind.STATE.equals(channel.getKind()) && isLinked(channel.getUID()));
     }
 
-    default Optional<CommonInterface> getHomeHandler() {
-        CommonInterface bridgeHandler = getBridgeHandler();
-        if (bridgeHandler != null) {
-            return bridgeHandler.getCapabilities().get(HomeCapability.class).isPresent() ? Optional.of(bridgeHandler)
-                    : Optional.empty();
-        }
-        return Optional.empty();
+    default Optional<CommonInterface> recurseUpToHomeHandler(@Nullable CommonInterface handler) {
+        return handler == null ? Optional.empty()
+                : handler.getCapabilities().get(HomeCapability.class).isPresent() ? Optional.of(handler)
+                        : recurseUpToHomeHandler(handler.getBridgeHandler());
     }
 
     default List<CommonInterface> getActiveChildren() {
@@ -146,7 +143,8 @@ public interface CommonInterface {
     }
 
     default <T extends RestCapability<?>> Optional<T> getHomeCapability(Class<T> clazz) {
-        return getHomeHandler().map(handler -> handler.getCapabilities().get(clazz)).orElse(Optional.empty());
+        return recurseUpToHomeHandler(this).map(handler -> handler.getCapabilities().get(clazz))
+                .orElse(Optional.empty());
     }
 
     default void setNewData(NAObject newData) {

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/AlarmEventCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/AlarmEventCapability.java
@@ -36,7 +36,7 @@ public class AlarmEventCapability extends HomeSecurityThingCapability {
 
     @Override
     public List<NAObject> updateReadings() {
-        return securityCapability.map(cap -> cap.getDeviceLastEvent(handler.getId(), moduleType.apiName))
+        return getSecurityCapability().map(cap -> cap.getDeviceLastEvent(handler.getId(), moduleType.apiName))
                 .map(event -> List.of((NAObject) event)).orElse(List.of());
     }
 }

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/CameraCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/CameraCapability.java
@@ -70,7 +70,7 @@ public class CameraCapability extends HomeSecurityThingCapability {
         String newVpnUrl = newData.getVpnUrl();
         if (newVpnUrl != null && !newVpnUrl.equals(vpnUrl)) {
             // This will also decrease the number of requests emitted toward Netatmo API.
-            localUrl = newData.isLocal() ? securityCapability.map(cap -> cap.ping(newVpnUrl)).orElse(null) : null;
+            localUrl = newData.isLocal() ? getSecurityCapability().map(cap -> cap.ping(newVpnUrl)).orElse(null) : null;
             cameraHelper.setUrls(newVpnUrl, localUrl);
             eventHelper.setUrls(newVpnUrl, localUrl);
         }
@@ -114,7 +114,7 @@ public class CameraCapability extends HomeSecurityThingCapability {
     @Override
     public void handleCommand(String channelName, Command command) {
         if (command instanceof OnOffType && CHANNEL_MONITORING.equals(channelName)) {
-            securityCapability.ifPresent(cap -> cap.changeStatus(localUrl, OnOffType.ON.equals(command)));
+            getSecurityCapability().ifPresent(cap -> cap.changeStatus(localUrl, OnOffType.ON.equals(command)));
         } else {
             super.handleCommand(channelName, command);
         }
@@ -123,7 +123,7 @@ public class CameraCapability extends HomeSecurityThingCapability {
     @Override
     protected void beforeNewData() {
         super.beforeNewData();
-        securityCapability.ifPresent(cap -> {
+        getSecurityCapability().ifPresent(cap -> {
             NAObjectMap<HomeDataPerson> persons = cap.getPersons();
             descriptionProvider.setStateOptions(personChannelUID, persons.values().stream()
                     .map(p -> new StateOption(p.getId(), p.getName())).collect(Collectors.toList()));
@@ -133,7 +133,7 @@ public class CameraCapability extends HomeSecurityThingCapability {
     @Override
     public List<NAObject> updateReadings() {
         List<NAObject> result = new ArrayList<>();
-        securityCapability.ifPresent(cap -> {
+        getSecurityCapability().ifPresent(cap -> {
             HomeEvent event = cap.getDeviceLastEvent(handler.getId(), moduleType.apiName);
             if (event != null) {
                 result.add(event);

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/CameraCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/CameraCapability.java
@@ -14,7 +14,6 @@ package org.openhab.binding.netatmo.internal.handler.capability;
 
 import static org.openhab.binding.netatmo.internal.NetatmoBindingConstants.*;
 import static org.openhab.binding.netatmo.internal.utils.ChannelTypeUtils.*;
-import static org.openhab.binding.netatmo.internal.utils.ChannelTypeUtils.toStringType;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -107,7 +106,8 @@ public class CameraCapability extends HomeSecurityThingCapability {
         // The channel should get triggered at last (after super and sub methods), because this allows rules to access
         // the new updated data from the other channels.
         final String eventType = event.getEventType().name();
-        handler.getHomeHandler().ifPresent(homeHandler -> homeHandler.triggerChannel(CHANNEL_HOME_EVENT, eventType));
+        handler.recurseUpToHomeHandler(handler)
+                .ifPresent(homeHandler -> homeHandler.triggerChannel(CHANNEL_HOME_EVENT, eventType));
         handler.triggerChannel(CHANNEL_HOME_EVENT, eventType);
     }
 

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeCapability.java
@@ -47,7 +47,7 @@ public class HomeCapability extends RestCapability<HomeApi> {
     private final Logger logger = LoggerFactory.getLogger(HomeCapability.class);
     private final Set<FeatureArea> featureAreas = new HashSet<>();
     private final NetatmoDescriptionProvider descriptionProvider;
-    private Set<String> homeIds = Set.of();
+    private final Set<String> homeIds = new HashSet<>();
 
     public HomeCapability(CommonInterface handler, NetatmoDescriptionProvider descriptionProvider) {
         super(handler, HomeApi.class);
@@ -58,7 +58,13 @@ public class HomeCapability extends RestCapability<HomeApi> {
     public void initialize() {
         super.initialize();
         HomeConfiguration config = handler.getConfiguration().as(HomeConfiguration.class);
-        homeIds = Set.of(config.getId(), config.energyId, config.securityId);
+        homeIds.add(config.getId());
+        if (!config.energyId.isBlank()) {
+            homeIds.add(config.energyId);
+        }
+        if (!config.securityId.isBlank()) {
+            homeIds.add(config.securityId);
+        }
     }
 
     @Override

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeCapability.java
@@ -68,6 +68,12 @@ public class HomeCapability extends RestCapability<HomeApi> {
     }
 
     @Override
+    public void dispose() {
+        homeIds.clear();
+        super.dispose();
+    }
+
+    @Override
     protected void updateHomeData(HomeData home) {
         if (hasArea(FeatureArea.SECURITY) && !handler.getCapabilities().containsKey(SecurityCapability.class)) {
             handler.getCapabilities().put(new SecurityCapability(handler));

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeSecurityThingCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/HomeSecurityThingCapability.java
@@ -36,8 +36,8 @@ public class HomeSecurityThingCapability extends Capability {
     protected final EventChannelHelper eventHelper;
 
     private Optional<WebhookServlet> webhook = Optional.empty();
-    protected Optional<SecurityCapability> securityCapability = Optional.empty();
-    protected Optional<HomeCapability> homeCapability = Optional.empty();
+    private Optional<SecurityCapability> securityCapability = Optional.empty();
+    private Optional<HomeCapability> homeCapability = Optional.empty();
 
     public HomeSecurityThingCapability(CommonInterface handler, NetatmoDescriptionProvider descriptionProvider,
             List<ChannelHelper> channelHelpers) {
@@ -49,16 +49,23 @@ public class HomeSecurityThingCapability extends Capability {
         eventHelper.setModuleType(moduleType);
     }
 
-    @Override
-    public void initialize() {
-        super.initialize();
-        securityCapability = handler.getHomeCapability(SecurityCapability.class);
-        homeCapability = handler.getHomeCapability(HomeCapability.class);
-        ApiBridgeHandler accountHandler = handler.getAccountHandler();
-        if (accountHandler != null) {
-            webhook = accountHandler.getWebHookServlet();
-            webhook.ifPresent(servlet -> servlet.registerDataListener(handler.getId(), this));
+    protected Optional<SecurityCapability> getSecurityCapability() {
+        if (securityCapability.isEmpty()) {
+            securityCapability = handler.getHomeCapability(SecurityCapability.class);
+            ApiBridgeHandler accountHandler = handler.getAccountHandler();
+            if (accountHandler != null) {
+                webhook = accountHandler.getWebHookServlet();
+                webhook.ifPresent(servlet -> servlet.registerDataListener(handler.getId(), this));
+            }
         }
+        return securityCapability;
+    }
+
+    protected Optional<HomeCapability> getHomeCapability() {
+        if (homeCapability.isEmpty()) {
+            homeCapability = handler.getHomeCapability(HomeCapability.class);
+        }
+        return homeCapability;
     }
 
     @Override

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/PersonCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/PersonCapability.java
@@ -56,7 +56,7 @@ public class PersonCapability extends HomeSecurityThingCapability {
     @Override
     protected void beforeNewData() {
         super.beforeNewData();
-        securityCapability.ifPresent(cap -> {
+        getSecurityCapability().ifPresent(cap -> {
             Stream<HomeDataModule> cameras = cap.getModules().values().stream()
                     .filter(module -> module.getType() == ModuleType.WELCOME);
             descriptionProvider.setStateOptions(cameraChannelUID,
@@ -67,7 +67,7 @@ public class PersonCapability extends HomeSecurityThingCapability {
     @Override
     public void handleCommand(String channelName, Command command) {
         if ((command instanceof OnOffType) && CHANNEL_PERSON_AT_HOME.equals(channelName)) {
-            securityCapability.ifPresent(cap -> cap.setPersonAway(handler.getId(), OnOffType.OFF.equals(command)));
+            getSecurityCapability().ifPresent(cap -> cap.setPersonAway(handler.getId(), OnOffType.OFF.equals(command)));
         }
     }
 
@@ -88,7 +88,7 @@ public class PersonCapability extends HomeSecurityThingCapability {
     @Override
     public List<NAObject> updateReadings() {
         List<NAObject> result = new ArrayList<>();
-        securityCapability.ifPresent(cap -> {
+        getSecurityCapability().ifPresent(cap -> {
             HomeEvent event = cap.getLastPersonEvent(handler.getId());
             if (event != null) {
                 result.add(event);

--- a/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/PresenceCapability.java
+++ b/bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/capability/PresenceCapability.java
@@ -62,6 +62,6 @@ public class PresenceCapability extends CameraCapability {
     }
 
     private void changeFloodlightMode(FloodLightMode mode) {
-        securityCapability.ifPresent(cap -> cap.changeFloodlightMode(handler.getId(), mode));
+        getSecurityCapability().ifPresent(cap -> cap.changeFloodlightMode(handler.getId(), mode));
     }
 }

